File: examples/swift/ToolServer.swift
import Foundation
import smpp_codec_ffi

// Basic single-threaded blocking server for demonstration
let PORT: UInt16 = 2775
let HEADER_LEN = 16

// Command IDs
let CMD_BIND_RECEIVER: UInt32 = 0x00000001
let CMD_BIND_TRANSMITTER: UInt32 = 0x00000002
let CMD_SUBMIT_SM: UInt32 = 0x00000004
let CMD_UNBIND: UInt32 = 0x00000006
let CMD_BIND_TRANSCEIVER: UInt32 = 0x00000009

func runServer() {
    let socket = CFSocketCreate(kCFAllocatorDefault, PF_INET, SOCK_STREAM, IPPROTO_TCP, 0, nil, nil)
    // .. This is getting complicated to do robustly in a single script file without a framework.
    // Swift scripts for networking usually use Network.framework or Foundation URLSession, but creating a raw TCP server
    // is verbose. I will write a simplified mock server logic or a skeleton that assumes Network framework if available.
    // Actually, let's use a very simple BSD socket approach if possible or just print instructions.
    
    print("Swift ToolServer implementation requires simpler networking for a script context.")
    print("Please see ToolClient.swift for the client implementation.")
}

// NOTE: Implementing a full TCP server in a single Swift script is non-trivial compared to Python/Kotlin/Node.
// For now, providing the logic structure.

import Network

@available(macOS 10.14, *)
func startNetworkServer() {
    let listener = try! NWListener(using: .tcp, on: NWEndpoint.Port(integerLiteral: PORT))
    
    listener.newConnectionHandler = { connection in
        print("New connection")
        connection.start(queue: .main)
        handleConnection(connection)
    }
    
    listener.start(queue: .main)
    print("Server listening on port \(PORT)")
    dispatchMain()
}

@available(macOS 10.14, *)
func handleConnection(_ connection: NWConnection) {
    func receiveHeader() {
        connection.receive(minimumIncompleteLength: HEADER_LEN, maximumLength: HEADER_LEN) { data, _, isComplete, error in
            if let data = data, data.count == HEADER_LEN {
                 // Parse Header
                 let header = [UInt8](data)
                 let commandLen = UInt32(header[0]) << 24 | UInt32(header[1]) << 16 | UInt32(header[2]) << 8 | UInt32(header[3])
                 let commandId = UInt32(header[4]) << 24 | UInt32(header[5]) << 16 | UInt32(header[6]) << 8 | UInt32(header[7])
                 let sequenceNumber = UInt32(header[12]) << 24 | UInt32(header[13]) << 16 | UInt32(header[14]) << 8 | UInt32(header[15])
                 
                 print(String(format: "Header: Len=%d, ID=0x%08X", commandLen, commandId))
                 
                 let bodyLen = Int(commandLen) - HEADER_LEN
                 if bodyLen > 0 {
                     connection.receive(minimumIncompleteLength: bodyLen, maximumLength: bodyLen) { bodyData, _, _, _ in
                        if let bodyData = bodyData {
                             let fullPdu = data + bodyData
                             processPdu(connection, commandId, sequenceNumber, fullPdu)
                        }
                        receiveHeader() // Next PDU
                     }
                 } else {
                     processPdu(connection, commandId, sequenceNumber, data)
                     receiveHeader() // Next PDU
                 }
            } else {
                connection.cancel()
            }
        }
    }
    receiveHeader()
}

@available(macOS 10.14, *)
func processPdu(_ connection: NWConnection, _ commandId: UInt32, _ sequenceNumber: UInt32, _ data: Data) {
    var responseData: Data?
    
    if commandId == CMD_BIND_TRANSCEIVER {
        let req = try! decodeBindRequest(buffer: data)
        print("Received Bind: \(req.systemId)")
         let resp = BindResponse(
            sequenceNumber: sequenceNumber,
            commandId: commandId | 0x80000000,
            systemId: req.systemId,
            scInterfaceVersionTlv: nil
        )
        responseData = encodeBindResponse(response: resp)
    } else if commandId == CMD_SUBMIT_SM {
         let _ = try! decodeSubmitSmRequest(buffer: data)
         print("Received SubmitSm")
         let resp = SubmitSmResponse(
             sequenceNumber: sequenceNumber,
             messageId: "MsgID_12345"
         )
         responseData = encodeSubmitSmResponse(response: resp)
    } else if commandId == CMD_UNBIND {
        print("Received Unbind")
        let resp = UnbindResponse(sequenceNumber: sequenceNumber)
        responseData = encodeUnbindResponse(response: resp)
    }
    
    if let resp = responseData {
        connection.send(content: resp, completion: .contentProcessed({ _ in
            print("Sent Response")
        }))
    }
}

if #available(macOS 10.14, *) {
    startNetworkServer()
} else {
    print("Requires macOS 10.14+")
}

File: examples/python/tool_server.py
import socket
import struct
import smpp_codec_ffi

HOST = "127.0.0.1"
PORT = 2775
HEADER_LEN = 16

# Command IDs (Requests)
CMD_BIND_RECEIVER = 0x00000001
CMD_BIND_TRANSMITTER = 0x00000002
CMD_QUERY_SM = 0x00000003
CMD_SUBMIT_SM = 0x00000004
CMD_DELIVER_SM = 0x00000005
CMD_UNBIND = 0x00000006
CMD_REPLACE_SM = 0x00000007
CMD_CANCEL_SM = 0x00000008
CMD_BIND_TRANSCEIVER = 0x00000009
CMD_ENQUIRE_LINK = 0x00000015
CMD_SUBMIT_MULTI_SM = 0x00000021
CMD_ALERT_NOTIFICATION = 0x00000102
CMD_DATA_SM = 0x00000103
CMD_BROADCAST_SM = 0x00000111
CMD_QUERY_BROADCAST_SM = 0x00000112
CMD_CANCEL_BROADCAST_SM = 0x00000113

def handle_client(conn, addr):
    print(f"Connected by {addr}")
    try:
        while True:
            header = conn.recv(HEADER_LEN)
            if not header or len(header) < HEADER_LEN:
                break
            
            command_len = struct.unpack(">I", header[0:4])[0]
            command_id = struct.unpack(">I", header[4:8])[0]
            command_status = struct.unpack(">I", header[8:12])[0]
            sequence_number = struct.unpack(">I", header[12:16])[0]
            
            print(f"Header: Len={command_len}, ID=0x{command_id:08X}, Status={command_status}, Seq={sequence_number}")
            
            body_len = command_len - HEADER_LEN
            body = b""
            if body_len > 0:
                body = conn.recv(body_len)
            
            full_pdu = header + body
            
            response_pdu = None
            
            if command_id in [CMD_BIND_RECEIVER, CMD_BIND_TRANSMITTER, CMD_BIND_TRANSCEIVER]:
                req = smpp_codec_ffi.decode_bind_request(full_pdu)
                print(f"Received Bind: {req.system_id}")
                resp = smpp_codec_ffi.BindResponse(
                     sequence_number=sequence_number,
                     command_id=command_id | 0x80000000,
                     system_id=req.system_id,
                     sc_interface_version_tlv=None
                )
                response_pdu = smpp_codec_ffi.encode_bind_response(resp)
                
            elif command_id == CMD_SUBMIT_SM:
                req = smpp_codec_ffi.decode_submit_sm_request(full_pdu)
                print(f"Received SubmitSm: {req.source_addr} -> {req.destination_addr}")
                
                # Concatenation Detection
                is_udh = (req.esm_class & 0x40) == 0x40
                sar_msg_ref = None
                sar_total = None
                sar_seq = None
                
                for tlv in req.tlvs:
                    if tlv.tag == smpp_codec_ffi.Tags.SAR_MSG_REF_NUM:
                        # Extract u16 from value (2 bytes, big endian)
                        sar_msg_ref = struct.unpack(">H", bytes(tlv.value))[0]
                    elif tlv.tag == smpp_codec_ffi.Tags.SAR_TOTAL_SEGMENTS:
                        sar_total = tlv.value[0]
                    elif tlv.tag == smpp_codec_ffi.Tags.SAR_SEGMENT_SEQNUM:
                        sar_seq = tlv.value[0]
                
                if is_udh:
                    print("  [DETECTION] UDH Concatenation detected")
                    # In a real server, we would parse the UDH header from req.short_message[0:6] or [0:7]
                    if len(req.short_message) > 6:
                         print(f"  [DEBUG] UDH Info (Raw): {req.short_message[:6].hex()}")
                
                if sar_msg_ref is not None:
                    print(f"  [DETECTION] SAR Concatenation detected: Ref={sar_msg_ref}, Part={sar_seq}/{sar_total}")

                resp = smpp_codec_ffi.SubmitSmResponse(
                    sequence_number=sequence_number,
                    message_id="MsgID_12345"
                )
                response_pdu = smpp_codec_ffi.encode_submit_sm_response(resp)

            elif command_id == CMD_DELIVER_SM:
                req = smpp_codec_ffi.decode_deliver_sm(full_pdu)
                print(f"Received DeliverSm: {req.source_addr}")
                resp = smpp_codec_ffi.DeliverSmResponse(
                    sequence_number=sequence_number,
                    command_status=0,
                    message_id="MsgID_DLR",
                    status_description=""
                )
                response_pdu = smpp_codec_ffi.encode_delivery_sm_response(resp)

            elif command_id == CMD_BROADCAST_SM:
                req = smpp_codec_ffi.decode_broadcast_sm_request(full_pdu)
                print(f"Received BroadcastSm")
                resp = smpp_codec_ffi.BroadcastSmResponse(
                    sequence_number=sequence_number,
                    message_id="BcastID_999",
                    optional_params=[]
                )
                response_pdu = smpp_codec_ffi.encode_broadcast_sm_response(resp)

            elif command_id == CMD_SUBMIT_MULTI_SM:
                req = smpp_codec_ffi.decode_submit_multi_request(full_pdu)
                print(f"Received SubmitMulti")
                resp = smpp_codec_ffi.SubmitMultiResponse(
                    sequence_number=sequence_number,
                    message_id="MsgID_Multi",
                    unsuccess_smes=[]
                )
                response_pdu = smpp_codec_ffi.encode_submit_multi_response(resp)

            elif command_id == CMD_QUERY_SM:
                req = smpp_codec_ffi.decode_query_sm_request(full_pdu)
                print(f"Received QuerySm")
                resp = smpp_codec_ffi.QuerySmResponse(
                    sequence_number=sequence_number,
                    message_id=req.message_id,
                    final_date="220101000000000R",
                    message_state=2, # Delivered
                    error_code=0
                )
                response_pdu = smpp_codec_ffi.encode_query_sm_response(resp)

            elif command_id == CMD_CANCEL_SM:
                req = smpp_codec_ffi.decode_cancel_sm_request(full_pdu)
                print(f"Received CancelSm")
                resp = smpp_codec_ffi.CancelSmResponse(
                    sequence_number=sequence_number
                )
                response_pdu = smpp_codec_ffi.encode_cancel_sm_response(resp)

            elif command_id == CMD_REPLACE_SM:
                req = smpp_codec_ffi.decode_replace_sm_request(full_pdu)
                print(f"Received ReplaceSm")
                resp = smpp_codec_ffi.ReplaceSmResponse(
                    sequence_number=sequence_number
                )
                response_pdu = smpp_codec_ffi.encode_replace_sm_response(resp)

            elif command_id == CMD_DATA_SM:
                req = smpp_codec_ffi.decode_data_sm(full_pdu)
                print(f"Received DataSm")
                resp = smpp_codec_ffi.DataSmResponse(
                    sequence_number=sequence_number,
                    command_status=0,
                    message_id="MsgID_Data",
                    status_description="",
                    optional_params=[]
                )
                response_pdu = smpp_codec_ffi.encode_data_sm_response(resp)

            elif command_id == CMD_ALERT_NOTIFICATION:
                req = smpp_codec_ffi.decode_alert_notification(full_pdu)
                print(f"Received AlertNotification")
                # No response

            elif command_id == CMD_QUERY_BROADCAST_SM:
                req = smpp_codec_ffi.decode_query_broadcast_sm_request(full_pdu)
                print(f"Received QueryBroadcastSm")
                resp = smpp_codec_ffi.QueryBroadcastSmResponse(
                    sequence_number=sequence_number,
                    message_id=req.message_id,
                    optional_params=[]
                )
                response_pdu = smpp_codec_ffi.encode_query_broadcast_sm_response(resp)

            elif command_id == CMD_CANCEL_BROADCAST_SM:
                req = smpp_codec_ffi.decode_cancel_broadcast_sm_request(full_pdu)
                print(f"Received CancelBroadcastSm")
                resp = smpp_codec_ffi.CancelBroadcastSmResponse(
                    sequence_number=sequence_number
                )
                response_pdu = smpp_codec_ffi.encode_cancel_broadcast_sm_response(resp)

            elif command_id == CMD_UNBIND:
                req = smpp_codec_ffi.decode_unbind_request(full_pdu)
                print(f"Received Unbind")
                resp = smpp_codec_ffi.UnbindResponse(
                    sequence_number=sequence_number
                )
                response_pdu = smpp_codec_ffi.encode_unbind_response(resp)
                if response_pdu:
                    conn.sendall(bytes(response_pdu))
                    print("Sent Unbind Response")
                break

            elif command_id == CMD_ENQUIRE_LINK:
                # Assuming EnquireLink isn't explicitly in FFI top-level decode yet, but it might be
                # If not available, we can just echo success
                print("Received EnquireLink")
                # Need to check if EnquireLink is exposed. If not, manual construct? 
                # Let's assume it is or just skip for now as not in main list above
                # But to be safe, standard response:
                resp_header = struct.pack(">IIII", 16, 0x80000015, 0, sequence_number)
                conn.sendall(resp_header)
                continue

            else:
                print(f"Unknown Command ID: 0x{command_id:08X}")

            if response_pdu:
                conn.sendall(bytes(response_pdu))
                print("Sent Response")

    finally:
        conn.close()

def main():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind((HOST, PORT))
        s.listen()
        print(f"Server listening on {HOST}:{PORT}")
        while True:
            conn, addr = s.accept()
            handle_client(conn, addr)

if __name__ == "__main__":
    main()
